src/h5web/vis-packs/core/complex/hooks.ts

```diff
@@ -0,0 +1,74 @@
+import ndarray from 'ndarray';
+import { useMemo } from 'react';
+import type { DimensionMapping } from '../../../dimension-mapper/models';
+import type { H5WebComplex } from '../../../providers/models';
+import { useMappedArray } from '../hooks';
+import { Bounds, Domain, ScaleType } from '../models';
+import { getDomainValidForScale, getNewBounds } from '../utils';
+
+export function usePhaseAmplitude(
+  value: H5WebComplex[],
+  dims: number[],
+  dimMapping: DimensionMapping,
+  scaleType: ScaleType = ScaleType.Linear
+): {
+  phaseArray: ndarray;
+  phaseDomain: Domain | undefined;
+  amplitudeArray: ndarray;
+  amplitudeDomain: Domain | undefined;
+} {
+  const [mappedArray] = useMappedArray(value, dims, dimMapping);
+
+  const [
+    phaseArray,
+    phaseBounds,
+    amplitudeArray,
+    amplitudeBounds,
+  ] = useMemo(() => {
+    const phaseValues: number[] = [];
+    const amplitudeValues: number[] = [];
+
+    let phaseBounds: Bounds = {
+      min: Infinity,
+      max: -Infinity,
+      positiveMin: Infinity,
+    };
+    let amplitudeBounds: Bounds = {
+      min: Infinity,
+      max: -Infinity,
+      positiveMin: Infinity,
+    };
+
+    (mappedArray.data as H5WebComplex[]).forEach(([real, imag]) => {
+      const phase = Math.atan2(imag, real);
+      phaseBounds = getNewBounds(phaseBounds, phase);
+      phaseValues.push(phase);
+
+      const amplitude = Math.sqrt(real ** 2 + imag ** 2);
+      amplitudeBounds = getNewBounds(amplitudeBounds, amplitude);
+      amplitudeValues.push(amplitude);
+    });
```

Comment:
A `reduce` would have been nice 😁

Reply:
I pushed a version with `reduce`: I am not fully convinced but tell me what you think ^^